add docs for webhook signature (#2591)
FusionAuth/fusionauth-site: site/docs/v1/tech/events-webhooks/signing.adoc

Author: old habits die hard

---
layout: doc
title: Signing Webhooks
description: Learn how to sign your Webhook events so you can verify they were sent by FusionAuth
navcategory: developer
---

== Signing Webhooks

> Signing Webhook events allows you to verify each event was sent by FusionAuth.  Webhook signatures provide an extra layer of security in addition to the methods described in link:/docs/v1/tech/events-webhooks/securing[Securing a Webhook].  Signature verification provides protection against bad actors spoofing webhook event payloads to look like they came from FusionAuth.

Reviewer: 🦅 👁️ 
```suggestion
Signing Webhook events allows you to verify each event was sent by FusionAuth.  Webhook signatures provide an extra layer of security in addition to the methods described in link:/docs/v1/tech/events-webhooks/securing[Securing a Webhook]. Signature verification provides protection against bad actors spoofing webhook event payloads to look like they came from FusionAuth.
```

This document covers configuring webhook signatures, verifying signatures in your webhook listener, and key rotation.

* <<Configuration>>
* <<Signature Verification>>
* <<Example Webhook Listener Code>>
* <<Testing>>
* <<Key Rotation>>

=== Configuration

Configuring webhook signatures in FusionAuth consists of generating a key and configuring your webhook to sign using that key.

image::webhook-key-gen.png[Webhook Key Generation,width=1200,role=shadowed]

Keys are generated or imported from [breadcrumb]#Settings -> Key Master#. Webhooks can be signed with three types of keys

* EC key - strongest cryptography, public key can be available
* RSA key - strong cryptography, public key can be available
* HMAC key - fast cryptography, requires manual key distribution

EC and RSA keys allow you to make public keys available through the `/.well-known/jwks.json` endpoint, which facilitates key rotation. If your webhook listener cannot make outbound network connections or you prefer to manually configure your key in your webhook listener, HMAC keys are a good option.

> For this example, we'll generate an RSA key pair.

Reviewer: There aren't details on generating the RSA key pair. It seems like it just jumps from "let's create an RSA key pair" straight to assigning it.

Author: I don't think additional details are really necessary in this guide, but added a link to the Key Master Guide.

> Next, you configure your webhook to sign the event with this key. From [breadcrumb]#Settings -> Webhooks#, click on the Edit button for your webhook (or create a new webhook). Select the Signature tab panel. Once you select the check box to enable webhook signatures, a select box allows you to choose the generated key.

Reviewer: Based on some other examples, we use `breadcrumb` when referring to tabs in the UI. I would recommend changing up how this refers to UI elements because the enable is not a checkbox, and the select is a dropdown ("select box" sounds to me like the redirect URI fields that you can enter and add multiple values).

```suggestion
Next, you configure your webhook to sign the event with this key. From [breadcrumb]#Settings -> Webhooks#, click on the Edit button for your webhook (or create a new webhook). Select the [breadcrumb]#Signature# tab panel. Once you enable webhook signatures, a select dropdown allows you to choose the generated key.
```

.Example webhook HTTP header
----
X-FusionAuth-Signature-JWT: eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Il9IMDd3VkcxZlYzbDVpaDc0ck54SUMzbmV2RSJ9.eyJyZXF1ZXN0X2JvZHlfc2hhMjU2IjoiS2VWKy9IR29JUXJ4dUU1WVBDUlI2QXVRT0p2ZWxkWU5OaGJWaTFpMjJxaz0ifQ.J70gqZVuTej8FfriQqJJZecCT6XOZKH6h6Te2ir_yrSwR3luhoj_R1vAZULdrktaFPqXFXbnq9prN8j3ddelUVA5SU51J-MWVhz1bkimLo8EEdJ47ytI_97rPqVK1YJ6FSiS8_o37gablaQZv2WDbZ6ap-t4hNU5m7uwZTW9DerKg9iQjMDUIlfafEwsROLfNPfK49IsCzBNCQ8SsinVbGU0dNbs9YfMAxNzSuEKdZOIXkRNgjPfWpPnkwBbroWUrrpcoAcBSQIYFajKV-MFRISnFZ_blYps16f95iQsuTfqBkBH3r59R5tFBP66FA1bvQJZVlAHJfdNTXnXx2F2BQ
----

The JWT decodes with:
[source,json]
.JWT header
----
{
   "alg": "RS256",
   "typ": "JWT",
   "kid": "_H07wVG1fV3l5ih74rNxIC3nevE"
 }
----

[source,json]
.JWT payload
----
{
  "request_body_sha256": "KeV+/HGoIQrxuE5YPCRR6AuQOJveldYNNhbVi1i22qk="
}
----

The `kid` identifies the Id of the key used to sign the JWT. JWT libraries can look the key up from the JWKS endpoint, or a locally stored key can be used. After verifying the JWT signature, the JWT's `request_body_sha256` payload claim is compared against your own calculated SHA-256 hash of the event body

=== Example Webhook Listener Code

> The following code (https://github.com/FusionAuth/fusionauth-example-javascript-webhooks/blob/main/signature-verify/app.js[available on Github]) demonstrates webhook signature verification with a simple Node server.

Reviewer: Not sure why this phrasing doesn't bother me here. Just confirming that you know this will display `(available on GitHub)` in the parentheses.

Author: Yep.  (and agree it was a bit awkward in the other place)

For this example, we'll generate an RSA public key pair.

Next, you configure your webhook to sign the event with this key.  From [breadcrumb]#Settings -> Webhooks#, click on the Edit button for your webhook (or create a new webhook).  Select the Signature tab panel.  Once you select the check box to enable webhook signatures, a select box allows you to choose the generated key.

image::webhook-settings-signature.png[Webhook Settings Signature,width=1200,role=shadowed]

=== Signature Verification

The webhook signature is provided in the HTTP header `X-FusionAuth-Signature-JWT` as a signed JWT with a claim of `request_body_sha256` containing the sha256 hash of the webhook event payload.

Your webhook listener can verify the signature by

- Verify the JWT is properly signed
- Decode the JWT
- Compare the JWT's `request_body_sha256` claim against your own calculated sha256 hash of the event body

> [source,ini]

Author: it's genericish config file formatting.  Renders better than `text`.

----

The JWT decodes with:
[source,json]
.JWT header
----
{
   "alg": "RS256",
   "typ": "JWT",
   "kid": "_H07wVG1fV3l5ih74rNxIC3nevE"
 }
----

[source,json]
.JWT payload
----
{
  "request_body_sha256": "KeV+/HGoIQrxuE5YPCRR6AuQOJveldYNNhbVi1i22qk="
}
----

The `kid` identifies the Id of the key used to sign the JWT.  JWT libraries can look the key up from the JWKS endpoint, or a locally stored key can be used.  After verifying the JWT signature, the JWT's `request_body_sha256` payload claim is compared against your own calculated sha256 hash of the event body

=== Example Webhook Listener Code

The following code demonstrates webhook signature verification with a simple Node server.

[source,javascript]
.Example Node.js Webhook Signature Verifier
----

> const bodyParser = require('body-parser');

Reviewer: Sometimes for larger examples we put the JavaScript in an example repository and then include it from the raw GitHub URL like some of the WebAuthn stuff:
```
https://raw.githubusercontent.com/FusionAuth/fusionauth-example-javascript-webauthn/main/buffer-to-base64url.js
```

Not sure whether we want to do that with this one or not. I see that there is other JS for writing a webhook that is pretty lengthy.

Author: good question.

Reviewer 2: I would love to have this code in a separate repo, maybe even with tests?

Author: Done!  (without the tests)

  const hashPayload = req.rawBody;
  console.log("\n req.rawBody: " + hashPayload);

  const jwt = Buffer.from(req.get(signatureHeader) || '', 'utf8');

  try {
    const { payload, protectedHeader } = await jose.jwtVerify(jwt, cachedRemoteJWKS);

    const body_sha256 = crypto.createHash('sha256').update(hashPayload).digest('base64');

    // Compare digest signature with signature sent by provider
    if (payload.request_body_sha256 === body_sha256) {
      console.log("Valid signature");
      // Do your webhook event processing here
      res.json({ message: "Success" });
    } else {
      console.log("Invalid signature");
      // skip this event
      res.status(401).send('Unauthorized');
    }
  } catch (err) {
    console.log("Invalid JWT header");
    res.status(401).send('Unauthorized');
  }
});

app.listen(port, function () {
  console.log(`Example app listening on port ${port}!`);
});
----

=== Testing

The
link:/docs/v1/tech/events-webhooks/#test-a-webhook[Webhook Testing] page provides a quick way to test your webhook signature configuration and signature verification on your webhook listener.

=== Key Rotation

> link:/docs/v1/tech/tutorials/key-rotation[Rotating keys] regularly is an important part of a defense-in-depth strategy.  The type of key used for signing webhook events determines the process for rotating keys.

Reviewer: ```suggestion
link:/docs/v1/tech/tutorials/key-rotation[Rotating keys] regularly is an important part of a defense-in-depth strategy.  The type of key used for signing webhook events and the method used for fetching that key determines the process for rotating keys.
```

> * Keys with public keys (RSA or EC) *and* signature verification dynamically fetches public key from  `.well-known/jwks.json` endpoint

Reviewer: ```suggestion
* Signatures validated using a public key (RSA or EC) where signature verification dynamically fetches public key from  `.well-known/jwks.json` endpoint
```
